### PR TITLE
2681 s1 enhancements

### LINKED ISSUE #2681: TANF Section 1 validation clean-up
**Description:**

After #2369 is implemented, some of the validation checks for TANF Section 1 data need updates, either due to on-going revisions OFA is making based on STT feedback to the new coding instructions or because the pre-specified checks have not been implemented. These include:

- ~~`DATE_OF_BIRTH` in the T2 and T3 records are stored and validated as integers instead of date (YYYYMMDD)~~
    - Moved to #2746
- `WORK_PART_STATUS` values are `01,02,05,07,09,15,16,17,18,99`
- `if item 48 == 11 and AGE < 19, then item 38 != 1`
- `FIPS_CODE` still needs to be stored in T1  from the header record  (`state_fips`) 
- `FUNDING_STREAM` value options will revert back to only having 2 options. OFA has decided not to use value `3`. 
- ~~`GENDER` should allow for space-fill (T2 and T3)~~
- `OTHER_AMOUNT` and `TRANSITION_SERVICES_AMOUNT` should be excluded from sum amount cat 3 validator. 
- `DISPOSITION` only has one possible value now. anything that isnt `1` should be flagged as out of range. 

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] validation checks noted above are accounted for
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] _Task 1_
- [ ] _Task 2_
- [ ] _Task 3_
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

- @ADPennington to create an ssp section 1 enhancement ticket
- _Note 2_
- _Note 3_

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- [_TANF Sec 1 WorkBook_](https://gorafttech.sharepoint.com/sites/TDRSResearchDesign/Shared%20Documents/Forms/AllItems.aspx?id=%2Fsites%2FTDRSResearchDesign%2FShared%20Documents%2FDev%2Fdev%2FParsing%2Fdata%2Ftanf%2FSection1&viewid=981949f0%2D98e1%2D4ace%2D8da8%2D436aca592a46)
- _Doc 2_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- _Open Question 1_
- _Open Question 2_

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Updated WORK_PART_STATUS
- Added missing cat3 validator
- Updated FUNDING_STREAM validators
- Updated GENDER validators
- Updated OTHER_AMOUNT & TRANSITION_SERVICES_AMOUNT validators
- Updated DISPOSITION validators

Note: The FIPS/Tribe code update already exists because of the relation in the database. Accessing those fields within the record has also been covered in #2805 by adding STT as a queryable field in the search indices.

Pull request closes #2681 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
pytest -k test_parse.py
```

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] validation checks noted above are accounted for
- [x] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py [hunk 1]
```diff
@@ -112,7 +112,7 @@
                     condition_function=validators.oneOf((1, 2)),
                     result_field="WORK_PART_STATUS",
                     result_function=validators.oneOf(
-                        ["01", "02", "05", "07", "09", "15", "17", "18", "19", "99"]
+                        ["01", "02", "05", "07", "09", "15", "17", "18", "19", "  "]
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py]: i might be overthinking this, but... `WORK_PART_STATUS` is an optional field, so blanks are handled within `parse_and_validate`. i wonder if `if_then_validator` should do something similar, taking into account the `required` attribute of the field so blanks (or potentially other null/empty characters) don't explicitly need to be added to the postparsing validators

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 2]
```diff
@@ -598,3 +598,51 @@ def validate_header_rpt_month_year(datafile, header, generate_error):
...
+        RELATIONSHIP_HOH = int(RELATIONSHIP_HOH)
+
+        DOB = (
+            instance["DATE_OF_BIRTH"]
+            if type(instance) is dict
+            else getattr(instance, "DATE_OF_BIRTH")
+        )
+        DOB = str(DOB)
+        # 1974 01 14
+        DOB_datetime = datetime.datetime.strptime(DOB, '%Y%m%d')
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/parsers/validators.py]: could one of these operations throw an exception if the dob is blank?

File: tdrs-backend/tdpservice/settings/common.py [hunk 3]
```diff
@@ -194,7 +194,7 @@ class Common(Configuration):
 
     # Logging
     # set level as 'INFO' if env var is not set
-    LOGGING_LEVEL = os.getenv('LOGGING_LEVEL', 'INFO')
+    LOGGING_LEVEL = "WARNING"#os.getenv('LOGGING_LEVEL', 'INFO')
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/settings/common.py]: Remnant from debugging locally? Surely this wasn't in the ACs from original ticket.

**Review by andrew-jameson — APPROVED** [review]

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py (hunk 1 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py]: I think we could definitely do that and doing so prevents errors where we forgot to include the blank. Does that creep on scope for this ticket though? Im happy to implement here as long as no one objects.

File: tdrs-backend/tdpservice/parsers/validators.py (hunk 2 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/validators.py]: Absolutely haha. I will make this safer!

File: tdrs-backend/tdpservice/settings/common.py (hunk 3 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/settings/common.py]: Yep, debugging remnant. Will fix

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 4]
```diff
@@ -442,44 +442,44 @@ def validate(instance):
...
+                else getattr(instance, "MONTHS_FED_TIME_LIMIT")
+            )
+            if FAMILY_AFFILIATION == 1 and (RELATIONSHIP_HOH == 1 or RELATIONSHIP_HOH == 2):
+                if MONTHS_FED_TIME_LIMIT is None or int(MONTHS_FED_TIME_LIMIT) < 1:
+                    return false_case
+                else:
+                    return true_case
+            else:
+                return true_case
+        except Exception:
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/validators.py:538]: suggest adding debug log. Try to add log which includes the fields at least for more readable log

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 5]
```diff
@@ -488,50 +488,41 @@ def validate__FAM_AFF__HOH__Count_Fed_Time():
...
+                else getattr(instance, "COUNTABLE_MONTH_FED_TIME")
             )
+            if FAMILY_AFFILIATION == 1 and (RELATIONSHIP_HOH == 1 or RELATIONSHIP_HOH == 2):
+                if int(COUNTABLE_MONTH_FED_TIME) < 1:
+                    return false_case
+                else:
+                    return true_case
+            else:
+                return true_case
+        except Exception:
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/validators.py:586]: Nit: add debug log

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 6]
```diff
@@ -598,3 +589,51 @@ def validate_header_rpt_month_year(datafile, header, generate_error):
...
+            AGE = today.year - DOB_datetime.year - ((today.month, today.day) < (DOB_datetime.month, DOB_datetime.day))
+
+            if WORK_ELIGIBLE_INDICATOR == 11 and AGE < 19:
+                if RELATIONSHIP_HOH != 1:
+                    return false_case
+                else:
+                    return true_case
+            else:
+                return true_case
+        except Exception:
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/validators.py:712]: Add debug log

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py (hunk 1 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py]: captured in #2821

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Review by jtimpe — APPROVED** [review]

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py [hunk 7]
```diff
@@ -113,7 +113,7 @@
...
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py]:
```suggestion
                        ["01", "02", "05", "07", "09", "15", "17", "18", "19", "99"]
```

Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py]: this was my mistake.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py [hunk 8]
```diff
@@ -269,7 +270,7 @@
                     type="number",
                     startIndex=44,
                     endIndex=45,
-                    required=True,
+                    required=False,
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py]: please revert. my mistake.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py [hunk 9]
```diff
@@ -221,7 +221,7 @@
             type="number",
             startIndex=43,
             endIndex=44,
-            required=True,
+            required=False,
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py]:
```suggestion
            required=True,
```

my mistake.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py [hunk 10]
```diff
@@ -526,7 +526,7 @@
             type="number",
             startIndex=84,
             endIndex=85,
-            required=True,
+            required=False,
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py]:
```suggestion
            required=True,
```

**Comment by ADPennington** [comment]
temporarily adding `Blocked` to this PR until discussion with DIGIT team today.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t1.py [hunk 11]
```diff
@@ -209,7 +207,7 @@
                     endIndex=31,
                     required=True,
                     validators=[
-                        validators.oneOf([1, 2]),
+                        validators.matches(1)
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t1.py:214]: @elipe17 can we drop 3 as a valid possible option for item # 13 (`RECEIVES_SUB_HOUSING`)? line 258

**Review by atrimpe — APPROVED** [review]

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 12]
```diff
@@ -623,3 +625,57 @@ def validate_header_rpt_month_year(datafile, header, generate_error):
...
+            DOB = (
+                instance["DATE_OF_BIRTH"]
+                if type(instance) is dict
+                else getattr(instance, "DATE_OF_BIRTH")
+            )
+
+            DOB = str(DOB)
+            DOB_datetime = datetime.datetime.strptime(DOB, '%Y%m%d')
+            today = datetime.date.today()
+            AGE = today.year - DOB_datetime.year - ((today.month, today.day) < (DOB_datetime.month, DOB_datetime.day))
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/validators.py]: @elipe17 Age calculation is relative to the reporting period (`RPT_MONTH_YEAR`). per the workbook:

`requires calculating AGE: DOB (item #32) must be valid; convert RPT_MON_YEAR (item #4) to YYYYMMDD and set DD to 01; AGE formula: (rpt_month_year - DOB)/365.25`

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py [hunk 13]
```diff
@@ -504,7 +505,6 @@
                                 "17",
                                 "18",
                                 "19",
-                                "99",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py:510]: @elipe17 i think i mangled up the AC on this one. the following is the requirement:

- required field
- range of possible values is: `01,02,05,07,09,15,16,17,18,99`

**Review by ADPennington — CHANGES_REQUESTED** [review]
@elipe17 couple changes needed. otherwise, looks good!

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py (hunk 13 above)
Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py:510]: @elipe17 this still needs to be a required field.

Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py:510]: @ADPennington my apologies, apparently my brain wasnt working! Pushed the update for this.

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 14]
```diff
@@ -623,3 +625,64 @@ def validate_header_rpt_month_year(datafile, header, generate_error):
...
+                instance["RPT_MONTH_YEAR"]
+                if type(instance) is dict
+                else getattr(instance, "RPT_MONTH_YEAR")
+            )
+
+            RPT_MONTH_YEAR += "01"
+
+            DOB_datetime = datetime.datetime.strptime(DOB, '%Y%m%d')
+            RPT_MONTH_YEAR_datetime = datetime.datetime.strptime(DOB, '%Y%m%d')
+            AGE = (RPT_MONTH_YEAR_datetime - DOB_datetime).days / 365.25
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/validators.py:703]: @elipe17 i'm not seeing the error message for this one. 

2681p1

Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/validators.py:703]: @ADPennington, there was a type error in the validator. I have resolved and the error will generate now

**Review by atrimpe — APPROVED** [review]
👍

**Comment by ADPennington** [comment]
@elipe17 approval here is pending your input on the test notes [here](https://hackmd.io/P4sQkTRORLGjiodfgRwMtA#2681-s1-enhancements), specifically:  when AGE is invalid (i.e. DOB==`20 0114`), WEI == 11, and HOH == 1: the following errors are raised:
`20 0114 does not have exactly 8 digits` ✔️ 
`20 must be larger than year 1900` ✔️ 
`If WORK_ELIGIBLE_INDICATOR == 11 and AGE < 19, then RELATIONSHIP_HOH != 1`  ⚠️ (unsure why this would be raised since age is invalid)

**Comment by elipe17** [comment]
> @elipe17 approval here is pending your input on the test notes [here](https://hackmd.io/P4sQkTRORLGjiodfgRwMtA#2681-s1-enhancements), specifically: when AGE is invalid (i.e. DOB==`20 0114`), WEI == 11, and HOH == 1: the following errors are raised: `20 0114 does not have exactly 8 digits` ✔️ `20 must be larger than year 1900` ✔️ `If WORK_ELIGIBLE_INDICATOR == 11 and AGE < 19, then RELATIONSHIP_HOH != 1` ⚠️ (unsure why this would be raised since age is invalid)

@ADPennington that error arrises because of the exception handling in the new validator. If you would prefer to return a `true case` during an exception that is a quick change. Let me know your thoughts.

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 15]
```diff
@@ -412,8 +412,8 @@ def dateDayIsValid():
 def olderThan(min_age):
     """Validate that value is larger than min_age."""
     return make_validator(
-        lambda value: date.today().year - int(str(value)[:4]) > min_age,
-        lambda value: (f"{str(value)[:4]} must be less than or equal to {date.today().year - min_age} "
+        lambda value: datetime.date.today().year - int(str(value)[:4]) > min_age,
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/validators.py:415]: just noting for reference that this validator is not needed.

**Review by ADPennington — APPROVED** [review]
thanks for the quick fix @elipe17 👍🏾 

test notes updated [here](https://hackmd.io/P4sQkTRORLGjiodfgRwMtA?view#2681-s1-enhancements)

Also, I noticed some updates to the `olderThan` validator, but to my knowledge, that validator isn't needed since we dont calculate age relative to today's date.

**Pull request merged**